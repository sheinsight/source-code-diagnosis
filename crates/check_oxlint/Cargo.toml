[package]
name = "check_oxlint"
version = "0.1.0"
edition = "2024"
 

[dependencies]
anyhow = { workspace = true }
napi-derive = { workspace = true }
napi = { workspace = true }
utils = { path = "../utils" }
beans = { path = "../beans" }
insta = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
pathdiff = { workspace = true }

oxc_linter = { workspace = true }
oxc = { workspace = true }
