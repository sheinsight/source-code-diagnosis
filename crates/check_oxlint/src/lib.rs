use std::{fmt::Display, rc::Rc, sync::Arc};

use beans::{Location, Span};
use napi_derive::napi;
use oxc::diagnostics::Severity;
use oxc_linter::{FixKind, LinterBuilder, Oxlintrc};
use serde::Serialize;
use utils::{GlobErrorHandler, GlobSuccessHandler, glob_by_semantic};

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintLabelsSpan {
  pub offset: u32,
  pub length: u32,
}

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintLabelsResponse {
  pub span: CheckOxlintLabelsSpan,
  pub loc: Location,
}

#[napi(object)]
#[derive(Debug, Clone, Serialize)]
pub struct CheckOxlintResponse {
  pub file_name: String,
  pub help: String,
  pub url: String,
  pub severity: String,
  pub code: String,
  pub message: String,
  pub labels: Vec<CheckOxlintLabelsResponse>,
}

impl Display for CheckOxlintResponse {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", serde_json::to_string_pretty(self).unwrap())
  }
}

pub fn check_oxlint(
  oxlint_config: String,
  args: utils::GlobArgs,
) -> anyhow::Result<Vec<CheckOxlintResponse>> {
  let cx_args: Oxlintrc = serde_json::from_str(&oxlint_config)?;

  let linter = LinterBuilder::from_oxlintrc(true, cx_args.clone())
    .with_fix(FixKind::None)
    .build();

  let responses = glob_by_semantic(
    |GlobSuccessHandler {
       path,
       parse,
       semantic,
       relative_path,
       ..
     }| {
      let module_record = Arc::new(oxc_linter::ModuleRecord::new(
        &path,
        &parse.module_record,
        &semantic,
      ));
      let semantic = Rc::new(semantic);

      let source_text = semantic.source_text();

      let diagnostics = linter.run(&path, semantic, module_record);

      let responses = diagnostics
        .into_iter()
        .map(|d| {
          let error = d.error;

          let help = error.help.as_deref().unwrap_or_default().to_string();

          let url = error.url.as_deref().unwrap_or_default().to_string();

          let severity = match error.severity {
            Severity::Advice => "advice".to_string(),
            Severity::Warning => "warning".to_string(),
            Severity::Error => "error".to_string(),
          };

          let labels = error
            .labels
            .as_ref()
            .map(|v| {
              v.iter()
                .map(|l| {
                  let loc = Location::with_source(
                    &source_text,
                    Span {
                      start: l.offset() as u32,
                      end: l.offset() as u32 + l.len() as u32,
                    },
                  );

                  return CheckOxlintLabelsResponse {
                    span: CheckOxlintLabelsSpan {
                      offset: l.offset() as u32,
                      length: l.len() as u32,
                    },
                    loc,
                  };
                })
                .collect::<Vec<_>>()
            })
            .unwrap_or_default();

          return CheckOxlintResponse {
            file_name: relative_path.clone(),
            help,
            url,
            severity,
            code: error.code.to_string(),
            message: error.message.to_string(),
            labels: labels,
          };
        })
        .collect::<Vec<_>>();

      Some(responses)
    },
    |GlobErrorHandler { .. }| None,
    &args,
  )?;

  Ok(responses.into_iter().flatten().collect())
}
