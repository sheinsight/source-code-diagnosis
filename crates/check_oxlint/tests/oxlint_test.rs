use check_oxlint::check_oxlint;
use std::env::current_dir;

macro_rules! test_oxlint {
  ($name:ident, $rule:expr) => {
    #[test]
    fn $name() -> anyhow::Result<()> {
      let curr = current_dir()?.join("tests").join("features").join($rule);

      let js_args = utils::GlobJsArgs {
        pattern: None,
        ignore: None,
        cwd: curr.to_string_lossy().to_string(),
      };

      let args = utils::GlobArgs::from(js_args);
      let mut responses = check_oxlint(
        format!(
          r#"{{
        "rules":{{
          "{}":"error"
        }}
      }}"#,
          $rule
        ),
        args,
      )?;

      responses.sort_by(|a, b| a.file_name.cmp(&b.file_name));

      responses.iter().for_each(|response| {
        insta::assert_snapshot!(response.to_string());
      });

      Ok(())
    }
  };
}

test_oxlint!(test_debugger, "no-debugger");
test_oxlint!(test_console, "no-console");

test_oxlint!(test_constructor_super, "constructor-super");

test_oxlint!(test_for_direction, "for-direction");

test_oxlint!(test_getter_return, "getter-return");

test_oxlint!(test_no_async_promise_executor, "no-async-promise-executor");

test_oxlint!(
  test_no_constant_binary_expression,
  "no-constant-binary-expression"
);

test_oxlint!(test_no_case_declarations, "no-case-declarations");

test_oxlint!(test_no_class_assign, "no-class-assign");

test_oxlint!(test_no_compare_neg_zero, "no-compare-neg-zero");

test_oxlint!(test_no_cond_assign, "no-cond-assign");

test_oxlint!(test_no_const_assign, "no-const-assign");

test_oxlint!(test_no_constant_condition, "no-constant-condition");

test_oxlint!(test_no_control_regex, "no-control-regex");

test_oxlint!(test_no_delete_var, "no-delete-var");

test_oxlint!(test_no_dupe_class_members, "no-dupe-class-members");

test_oxlint!(test_no_dupe_else_if, "no-dupe-else-if");

test_oxlint!(test_no_dupe_keys, "no-dupe-keys");

test_oxlint!(test_no_duplicate_case, "no-duplicate-case");

test_oxlint!(test_no_new_for_builtins, "unicorn/new-for-builtins");
