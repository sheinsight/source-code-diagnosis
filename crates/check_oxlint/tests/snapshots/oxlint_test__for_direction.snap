---
source: crates/check_oxlint/tests/oxlint_test.rs
expression: response.to_string()
snapshot_kind: text
---
{
  "file_name": "for-direction.js",
  "help": "Use while loop for intended infinite loop",
  "url": "https://oxc.rs/docs/guide/usage/linter/rules/eslint/for-direction.html",
  "severity": "error",
  "code": "eslint(for-direction)",
  "message": "The update clause in this loop moves the variable in the wrong direction",
  "labels": [
    {
      "span": {
        "offset": 16,
        "length": 6
      },
      "loc": {
        "start": {
          "line": 1,
          "col": 17
        },
        "end": {
          "line": 1,
          "col": 23
        }
      }
    },
    {
      "span": {
        "offset": 24,
        "length": 3
      },
      "loc": {
        "start": {
          "line": 1,
          "col": 25
        },
        "end": {
          "line": 1,
          "col": 28
        }
      }
    }
  ]
}
